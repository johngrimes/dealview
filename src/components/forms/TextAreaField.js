// @flow

import React from 'react'

import type { FieldErrors } from '../../utils/FormValidation.js'

type Props = {
  name: string,
  value: string,
  label?: string,
  placeholder?: string,
  errors?: FieldErrors,
  forceErrorDisplay?: boolean,
  focus?: string,
  onChange?: (value: string) => void,
  onFocus?: (fieldName: string) => void
}

type State = {
  value: string,
  touched: boolean
}

class TextAreaField extends React.Component {
  props: Props
  state: State
  ref: HTMLTextAreaElement
  handleChange: (event: Event) => true
  handleFocus: (event: Event) => true

  constructor(props: Props) {
    super(props)
    this.state = {
      value: props.value,
      touched: false
    }

    this.handleChange = (event) => {
      const target = event.target
      if (target instanceof HTMLTextAreaElement) {
        this.setState({
          value: target.value,
          touched: true
        })
        if (this.props.onChange) this.props.onChange(target.value)
      }
      return true
    }

    this.handleFocus = (event) => {
      const target = event.target
      if (this.props.onFocus && target instanceof HTMLTextAreaElement) {
        this.props.onFocus(target.name)
      }
      return true
    }
  }

  setFocus() {
    if (this.props.focus === this.props.name) {
      this.ref.focus()
    }
  }
  componentDidMount() { this.setFocus() }
  componentDidUpdate() { this.setFocus() }

  render() {
    const labelTag = this.props.label ? <label htmlFor={this.props.name}>{this.props.label}</label> : null

    const errorTags = []
    if (this.props.errors) {
      this.props.errors.forEach((msg, i) =>
        errorTags.push(<div key={i} className='error'>{msg}</div>)
      )
    }

    const inputClass = this.state.touched && errorTags.length > 0
      ? 'with-errors'
      : ''

    return <div className='control-group'>
             {labelTag}
             <textarea name={this.props.name} className={inputClass} placeholder={this.props.placeholder} value={this.state.value} onChange={this.handleChange} onFocus={this.handleFocus} ref={input => this.ref = input} />
             {(this.props.forceErrorDisplay || this.state.touched) && errorTags.length > 0 && <div className='errors'>{errorTags}</div>}
           </div>
  }
}

export default TextAreaField
