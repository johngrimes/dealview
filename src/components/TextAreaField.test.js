import React from 'react'
import { shallow, mount } from 'enzyme'

import TextAreaField from './TextAreaField.js'

describe('TextAreaField', () => {
  it('should render', () => {
    const props = {
      name: 'someInput',
      value: 'something',
      label: 'Some Input',
      placeholder: 'Type here...',
    }
    const wrapper = shallow(<TextAreaField {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should be tolerant of an undefined value prop', () => {
    const props = {
      name: 'someInput',
      label: 'Some Input',
      placeholder: 'Type here...',
    }
    const wrapper = shallow(<TextAreaField {...props} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should display errors if touched', () => {
    const props = {
      name: 'someInput',
      errors: [ 'does not look right', 'smells funny' ],
    }
    const wrapper = shallow(<TextAreaField {...props} />)
    wrapper.setState(
      {
        value: 'something',
        touched: true,
      },
      () => {
        expect(wrapper.find('.error')).toHaveLength(2)
        expect(wrapper).toMatchSnapshot()
      }
    )
  })

  it('should not display errors if not touched', () => {
    const props = {
      name: 'someInput',
      type: 'text',
      errors: [ 'does not look right', 'smells funny' ],
    }
    const wrapper = shallow(<TextAreaField {...props} />)
    expect(wrapper.find('.error')).toHaveLength(0)
  })

  it('should display errors if not touched but forceErrorDisplay is true', () => {
    const props = {
      name: 'someInput',
      type: 'text',
      errors: [ 'does not look right', 'smells funny' ],
      forceErrorDisplay: true,
    }
    const wrapper = shallow(<TextAreaField {...props} />)
    expect(wrapper.find('.error')).toHaveLength(2)
  })

  it('should call onChange when the input changes', () => {
    const props = {
      name: 'someInput',
      type: 'text',
      onChange: jest.fn(),
    }
    const wrapper = shallow(<TextAreaField {...props} />)
    wrapper.find('textarea').prop('onChange')({
      target: { value: 'something' },
    })
    expect(props.onChange).toHaveBeenCalledWith('something')
  })

  it('should call onFocus when the input is focused', () => {
    const props = {
      name: 'someInput',
      type: 'text',
      onFocus: jest.fn(),
    }
    const wrapper = shallow(<TextAreaField {...props} />)
    wrapper.find('textarea').prop('onFocus')({ target: { name: 'someInput' } })
    expect(props.onFocus).toHaveBeenCalledWith('someInput')
  })

  it('should focus the input when the focus prop is passed', () => {
    const props = {
      name: 'someInput',
      type: 'text',
    }
    const wrapper = mount(<TextAreaField {...props} />)
    const spy = jest.spyOn(wrapper.find('textarea').getNode(), 'focus')
    wrapper.setProps({ focus: 'someInput' })
    expect(spy).toHaveBeenCalled()
  })
})
