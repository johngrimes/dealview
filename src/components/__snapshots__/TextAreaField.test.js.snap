// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAreaField should be tolerant of an undefined value prop 1`] = `
<div
  className="control-group"
>
  <label
    htmlFor="someInput"
  >
    Some Input
  </label>
  <textarea
    className=""
    name="someInput"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Type here..."
  />
</div>
`;

exports[`TextAreaField should display errors if touched 1`] = `
<div
  className="control-group"
>
  <textarea
    className=""
    name="someInput"
    onChange={[Function]}
    onFocus={[Function]}
  />
</div>
`;

exports[`TextAreaField should render 1`] = `
<div
  className="control-group"
>
  <label
    htmlFor="someInput"
  >
    Some Input
  </label>
  <textarea
    className=""
    name="someInput"
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="Type here..."
    value="something"
  />
</div>
`;
